### DISCUSSION TITLE
Create a Fedora CoreOS VM from an xz compressed cloud image

### DISCUSSION THREAD

**Discussion by webD97** [bpg/terraform-provider-proxmox#1212] [Q&A] [answered]
I'm currently using this provider (great work btw!) to set up a homelab k3s cluster. My nodes are running Fedora CoreOS and I'd like to bootstrap the VMs from the official qcow2 image.

I followed the [Create a VM from a Cloud Image](https://registry.terraform.io/providers/bpg/proxmox/latest/docs/guides/cloud-image) but unfortunately Proxmox does not seem to like the fact that CoreOS images are only available with xz compression (https://fedoraproject.org/coreos/download?stream=stable&arch=x86_64#baremetal), so this fails:

```tf
resource "proxmox_virtual_environment_download_file" "coreos_qcow2" {
  content_type = "iso"
  datastore_id = "local"
  node_name    = "pve"
  url          = "https://builds.coreos.fedoraproject.org/prod/streams/stable/builds/39.20240322.3.1/x86_64/fedora-coreos-39.20240322.3.1-qemu.x86_64.qcow2.xz"
}
```

> Could not download file 'fedora-coreos-39.20240322.3.1-qemu.x86_64.qcow2.xz', unexpected error: error download file by URL: received an HTTP 400 response - Reason: Parameter verification failed. (filename: wrong file extension)

I'm pretty sure this is a Proxmox limitation but maybe I am just missing something and someone has a better approach. My best idea so far is to create some local-exec magic that invokes coreos-downloader to create the uncompressed qcow2 locally, which is then used in a proxmox_virtual_environment_file, but that appears to be more a hack than a workaround.

Any ideas?

**Comment by bpg-dev** [comment]
Hi @webD97 👋🏼

I don't think Proxmox can use xz-compressed qcow2 disk images for VMs. Your idea about the local-exec approach is probably the best we can do to automate this atm.

**Reply by webD97** [reply]
I really hoped that I missed some feature - but it is as it is (at least for now). Thank you for you input!

**Comment by aaronkyriesenbach** [comment, marked as answer]
Howdy! Sorry to revive a dead thread but I wanted to chime in with my experience getting this working, namely that **Proxmox's ZST decompression is capable of decompressing XZ archives**. I have not found documentation on this anywhere, but have gotten this working successfully for the exact use case described here. Here is a complete, minimal working example (with bonus logic to pull the latest CoreOS image):
```
locals {
  metadata = jsondecode(data.http.coreos_stable_metadata.response_body)

  coreos_qemu_stable = local.metadata.architectures.x86_64.artifacts.qemu.formats["qcow2.xz"].disk
  download_url       = local.coreos_qemu_stable.location
  download_sum    = local.coreos_qemu_stable.sha256
}

data "http" "coreos_stable_metadata" {
  url = "https://builds.coreos.fedoraproject.org/streams/stable.json"
}

resource "proxmox_virtual_environment_download_file" "coreos_img" {
  content_type  = "iso"
  datastore_id   = "local"
  node_name    = "pve"

  url             = local.download_url
  checksum = local.download_sum

  checksum_algorithm      = "sha256"
  decompression_algorithm = "zst"

  file_name = "coreos-qemu.qcow2.xz.img"
}

# This is my personal VM module I use in my lab. It does not apply any specific customizations, so I am omitting the full contents of the module.
module "coreos-testing" {
  source = "./modules/vm"

  id       = xxx
  name = "coreos-tf"

  disk = {
    size    = 16 # The qcow2 image is 10Gb when pulled as of writing, but is successfully sized up during provisioning
    file_id = proxmox_virtual_environment_download_file.coreos_img.id
  }

  # I found that having less memory would cause boot failures without explicit mention as to why.
  # It took a while to figure out that this was because of the lack of memory and not because the disk extraction was going wrong.
  memory = 2

  agent_enabled = false

  cloud_init = {
    enabled = false
  }
}
```

**Reply by bpg-dev** [reply]
Good find @aaronkyriesenbach!

**Reply by agross** [reply]
I've been fighting all evening to create the VM from the extracted qcow2 image. Interestingly, it only works with `file_id`, whereas `import_from` always yields:

```
│ Error: error creating VM: All attempts fail:
│ #1: received an HTTP 400 response - Reason: Parameter verification failed. (scsi0: fcos:iso/fedora-coreos-42.20251012.3.0-proxmoxve.x86_64.qcow2.xz.img has wrong type 'iso' - needs to be 'images' or 'import')
│
│   with proxmox_virtual_environment_vm.vm,
│   on main.tf line 84, in resource "proxmox_virtual_environment_vm" "vm":
│   84: resource "proxmox_virtual_environment_vm" "vm" {
```

@bpg Is `import_from` a full replacement for the deprecated `file_id`?

**Reply by bpg-dev** [reply]
@agross could you paste your file resource config here?

**Reply by agross** [reply]
Sure, I think this is the relevant part:

```hcl
resource "proxmox_virtual_environment_download_file" "fcos_image" {
  # `import`, which is needed for `import_from`, would not support decompression.
  content_type = "iso"
  node_name    = var.pve_node_name
  datastore_id = var.pve_datastore_id

  url          = "https://builds.coreos.fedoraproject.org/prod/streams/${var.fcos_stream}/builds/${var.fcos_version}/x86_64/fedora-coreos-${var.fcos_version}-proxmoxve.x86_64.qcow2.xz"
  file_name    = "fedora-coreos-${var.fcos_version}-proxmoxve.x86_64.qcow2.xz.img"

  # https://github.com/bpg/terraform-provider-proxmox/discussions/1319#discussioncomment-14377866
  decompression_algorithm = "zst"
}

resource "proxmox_virtual_environment_vm" "vm" {
  node_name   = var.pve_node_name
  name        = var.hostname

  disk {
    datastore_id = var.vm_pve_datastore_id
    # Does not work with `import_from` because the file does not reside in the import dir.
    file_id      = resource.proxmox_virtual_environment_download_file.fcos_image.id
    size         = 10
    interface    = "scsi0"
  }
```

**Reply by bpg-dev** [reply]
Yes, you're right in the comment, that's a Proxmox's limitation.
`import_from` requires an image to be uploaded with `import` content type -- PVE does image conversion in the background.
And the `import` content type supports only raw / qcow2 images, not archives 🤷🏼

**Reply by agross** [reply]
Thank you for confirming! What do you think, does it make sense to un-deprecate `file_id` then?

**Reply by bpg-dev** [reply]
Well, it is not "officially" deprecated in the resource schema, just in the doc. We need better wording there. If you don't mind opening a quick PR to [update](https://github.com/bpg/terraform-provider-proxmox/blob/227ceef93c6e7fbe6476824f11dbc7b1f13e7575/docs/resources/virtual_environment_vm.md?plain=1#L299-L301) it? 🙏🏼

**Reply by agross** [reply]
In order to improve wording I need to know the reason why you decided to "semi-deprecate" it in the first place. As a mere user of the module (i.e. no idea about the Proxmox APIs used) it does not really make sense to have >1 way to define the VM's base image.

**Reply by bpg-dev** [reply]
Hey @agross, sorry, I missed your post.  
The reasons are mostly historical.

Support for the `import` content type is a recent feature, before that the only option for uploading VM images was `iso`. However, to use an uploaded image from the ISO storage as a VM disk it needs to be converted. The provider uses a "hack" to do that, it connects over SSH, runs a few CLI commands to handle the conversion, and then attaches the converted disk image to the VM. In order to do that it needs a reference to the image in the storage, hence the `file_id` attribute.

With the addition of `import` content type support in PVE 8.x we do not have to do that manual conversion anymore, PVE does it for us at VM creation time, but only if the source image for the disk is referenced in a special way through the `import-from` attribute.

There were a lot of complaints about the SSH requirement for VM creation, people asking "why?" or even "WTF, why?". So if you use an image via `import_from`, SSH is not used and not needed, and the SSH setup for the provider can be skipped altogether (unless you also need to upload snippets, that is the only remaining API gap where SSH hacks are used). This is the reason for the "soft deprecation", so if people are using old templates and examples they can update their configs to use import and ditch SSH.

Hope this makes sense, but let me know if you have more questions :)
